rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+
+-   **Naming** - variables and ops created inside modules are created within a
+    name scope so they can easily be tracked back to their creator.
+-   **Variable tracking** - variables referenced by a module (directly or via
+    dependencies on other modules) can be retrieved via a simple API.
+-   **Module tracking** - given a module a simple API exposes all modules that
+    this module depends on.
+
+Users are encouraged to subclass `tf.Module` to create their own stateful
+components and we expect and encourage an ecosystem of third party modules to
+evolve.
+
+## Motivation
+
+When defining neural networks it is convenient to describe the network as a DAG
+of connected components (e.g. datasets, layers, optimizers). In Python it is
+convenient to represent nodes in this DAG as objects which encapsulate node
+state (e.g. `tf.Variable`) and functionality (e.g. forward functions).
+
+We propose a lightweight "module" abstraction that serves as a base class for
+all types of node in this DAG (e.g. layers, optimizers, metrics) which is not
+tied to a particular higher level framework (e.g. Keras, Estimator or Sonnet).
+The module abstraction standardizes naming of variables and ops (to simplify
+debugging/visualisation) and dependency tracking (e.g. retrieving `tf.Variable`s
+and `tf.Module`s referenced by a given root module).
+
+Having a standard abstraction for stateful TensorFlow components will encourage
+users to create loosely coupled components which are easy to reason about in
+isolation while providing a standard mechanism for common tasks in a larger
+assembly (e.g. finding all trainable variables for a module tree given a root).
+
+We see such a lightweight abstraction as being particularly important for
+cutting edge research, where it is not possible for library/framework authors to
+reason about what their users will create. We believe that lightweight
+abstractions which can be extended or composed to build more complex systems are
+the way forward.
+
+## Background
+
+TensorFlow V1 has a number of abstractions for handling TensorFlow state (e.g.
+`tf.Variable`s). Broadly you can split them into "collections based" (e.g.
+`tf.variable_scope`, `tf.get_collection`, `tf.make_template`, `tf.layers.Layer`)
+and "Object Oriented" (e.g. `tf.keras.layers.Layer`).
+
+TensorFlow V2 is standardising on OO and `tf.keras` as the only stateful
+abstraction inside the default TensorFlow pip installation.
+
+Outside of TensorFlow there exist a number of third party libraries which also
+offer abstractions for managing TensorFlow state. For example the Sonnet library
+from DeepMind is built on top of `tf.make_template`.
+
+## Requirements
+
+### Functional
+
+**Naming `tf.Variable`s (and `tf.Operation`s)**
+
+-   Each module instance must have a name scope. This name should include the
+    name scope from when the module is constructed.
+-   Every method on the module must enter this name scope before calling user
```

Comment:
In eager op names are not a thing so that's good.

For `tf.function` names are uniqueified within the function body, so you would only get an _N suffix if there were actually two relu's within **the same** module (or you called your module multiple times sequentially).

As a nit, scope names are re-entrant (as in you can do `with tf.name_scope("foo/"): ...` as many times as you like) but op names need to be unique.

Example: https://gist.github.com/tomhennigan/a183a6570d2a6fb9bce08758ee2d78b6

Reply:
Thanks for the correction. I see now that this works, although I'm still a bit confused at where the _1 suffix comes from.
```
@tf.function
def f(x):
  name_scope = tf.name_scope('blah')
  with name_scope:
    with name_scope:
      return x + 1

g = f.get_concrete_function(tf.TensorSpec([1, 64], tf.float32)).graph
for node in g.as_graph_def().node:
  print(node.name)
  
...
x
blah_1/add/y
blah_1/add
Identity
```